readthedocs/doc_builder/python_environments.py

```diff
@@ -131,33 +131,32 @@ def is_obsolete(self):
                 environment_conf = json.load(fpath)
             env_python_version = environment_conf['python']['version']
             env_build_image = environment_conf['build']['image']
+            env_build_hash = environment_conf['build']['hash']
         except (IOError, TypeError, KeyError, ValueError):
             log.error('Unable to read/parse readthedocs-environment.json file')
             return False
 
-        # TODO: remove getattr when https://github.com/rtfd/readthedocs.org/pull/3339 got merged
-        build_image = getattr(self.config, 'build_image', self.version.project.container_image) or DOCKER_IMAGE  # noqa
-
+        build_image = self.config.build_image or DOCKER_IMAGE
         # If the user define the Python version just as a major version
         # (e.g. ``2`` or ``3``) we won't know exactly which exact version was
         # used to create the venv but we can still compare it against the new
         # one coming from the project version config.
         return any([
             env_python_version != self.config.python_full_version,
             env_build_image != build_image,
+            env_build_hash != self.build_env.image_hash,
```

Comment:
This will force wipe all existing builds, but perhaps that is a good thing. We'll have explicit metadata after the next build of all of these projects.

Reply:
Mmm...

This won't force the wipe because it will fail in the L134 (https://github.com/rtfd/readthedocs.org/pull/3793/files/05d74aec013e1b211798ba02e91636d331794a1d#diff-31f8391d8643c47c6caedf53930c0b7eR134) when trying to access the `build.hash` that will produce a KeyError exception and we will return `False`.

Then, the current build with the _old_ environment will be used but we will be saving the current _new_ attributes and... we have a bug here, right?